src/cf.data.pre

```diff
@@ -3057,6 +3057,8 @@ DEFAULT: none
 DOC_START
 	The OpenSSL engine to use. You will need to set this if you
 	would like to use hardware SSL acceleration for example.
+
+	Note: OpenSSL 3.0 and newer do not provide Engine support.
```

Comment:
The ENGINE API is not available in the libssl3 I am building/testing against. On top of that I have never seen this option used by anyone. So for now I have opted to deprecate the whole feature.

If anyone really wants to have an ENGINE with libssl3 they will need to refactor all the code making use of this config option and the variables it sets up.

Reply:
> The ENGINE API is not available in the libssl3 I am building/testing against.

The limitations specific to the library you use should not affect the decision regarding deprecation of a Squid feature that may be used by others. Perhaps you can rephrase the above statement in more general terms (e.g., OpenSSL v3 that comes by default with Ubuntu X disables ENGINE API support)? A _general_ statement about the necessary library availability could influence our decision to deprecate, of course.

> I have never seen this option used by anyone. So for now I have opted to deprecate the whole feature.

You have seen it used by bug [4112](https://bugs.squid-cache.org/show_bug.cgi?id=4112) reporter (at least) AFAICT.

> If anyone really wants to have an ENGINE with libssl3 they will need to refactor all the code making use of this config option and the variables it sets up.

The cost of support is a valid factor in our decision making process, of course. Can you give an example of Squid code that will need to be changed in order to preserve support for OpenSSL engine selection via `ssl_engine` directive in Squid built with OpenSSL v3? My superficial interpretation of OpenSSL v3 migration [documentation](https://www.openssl.org/docs/man3.0/man7/migration_guide.html) is that there should be no Squid changes as long as the OpenSSL library is built with that deprecated API/code enabled.

----

Assuming the continued support is indeed costly (which is to be determined above), one could address this change request by making the following changes:

1. Adjust PR description to document that we are deprecating OpenSSL Engine support (not because it is removed in OpenSSL v3 but) because it is deprecated in OpenSSL v3, popular distributions ship OpenSSL v3 without Engine support (if that is true), and we do not have the resources to update the corresponding Squid code to remain compatible with OpenSSL v3 (if the code needs updating).
2. Add deprecation warning (for OpenSSL before v3) to squid.conf parser or interpreter. You have already adjusted the code to fail with OpenSSL v3+.
3. Replace the current (incorrect) claim in cf.data.pre with something along these lines:

```suggestion
	Not supported in builds with OpenSSL v3 or newer.
```

There are other ways to address this change request, of course. The above is just an example that illustrates the missing pieces if we decide to deprecate `ssl_engine` support starting with OpenSSL v3 builds.

Reply:
Resolving as moved to https://github.com/squid-cache/squid/pull/694#discussion_r892425293